csharp/src/Ice/ColocatedStream.cs

```diff
@@ -12,85 +13,100 @@ namespace ZeroC.Ice
+                _socket.ReleaseFlowControlCredit(this);
             }
         }
 
-        protected override ValueTask<bool> ReceiveAsync(ArraySegment<byte> buffer, CancellationToken cancel) =>
+        protected override ValueTask<int> ReceiveAsync(Memory<byte> buffer, CancellationToken cancel) =>
             // This is never called because we override the default ReceiveFrameAsync implementation
             throw new NotImplementedException();
 
         protected override ValueTask ResetAsync(long errorCode) =>
             // A null frame indicates a stream reset.
             // TODO: Provide the error code?
-            _socket.SendFrameAsync(Id, frame: null, fin: true, CancellationToken.None);
+            _socket.SendFrameAsync(this, frame: null, fin: true, CancellationToken.None);
 
-        protected override ValueTask SendAsync(IList<ArraySegment<byte>> buffer, bool fin, CancellationToken cancel)
-        {
-            // This is only called for Initialize/GoAway frame on the control streams. Requests or responses are
-            // handled by the SendFrameAsync method override below.
-            if (!IsStarted)
-            {
-                Id = _socket.AllocateId(false);
-            }
-            Debug.Assert(IsControl && !IsBidirectional);
-            return _socket.SendFrameAsync(Id, buffer, fin, cancel);
-        }
+        protected override ValueTask SendAsync(IList<ArraySegment<byte>> buffer, bool fin, CancellationToken cancel) =>
+            _socket.SendFrameAsync(this, frame: buffer, fin: fin, cancel);
 
         /// <summary>Constructor for incoming colocated stream</summary>
-        internal ColocatedStream(long streamId, ColocatedSocket socket)
-            : base(streamId, socket) => _socket = socket;
+        internal ColocatedStream(ColocatedSocket socket, long streamId)
+            : base(socket, streamId) => _socket = socket;
 
         /// <summary>Constructor for outgoing colocated stream</summary>
-        internal ColocatedStream(bool bidirectional, ColocatedSocket socket)
-            : base(bidirectional, socket) => _socket = socket;
+        internal ColocatedStream(ColocatedSocket socket, bool isBidirectional, bool isControl)
+            : base(socket, isBidirectional, isControl) => _socket = socket;
 
-        internal void ReceivedFrame(object frame, bool fin) =>
-            // Run the continuation asynchronously if it's a response to ensure we don't end up calling user
-            // code which could end up blocking the AcceptStreamAsync task.
-            SignalCompletion((frame, fin), runContinuationAsynchronously: frame is OutgoingResponseFrame);
+        internal void ReceivedFrame(object frame, bool fin)
+        {
+            if (IsSignaled)
+            {
+                Debug.Assert(fin);
+                _streamable = frame;
+                if (!IsSignaled)
+                {
+                    SignalCompletion((frame, true));
+                }
+            }
```

Comment:
Added some comments:
```
                // The frame might already be signaled if the peer sends the request frame and then the request
                // stream data. SignaledSocketStream can't queue multiple signals so we have to keep track of the
                // second frame in _streamable here. This can only occur for the stream data and in this case
                // fin should be true since the data frame is the last frame that will be sent.
                //
                // TODO: All this code needs to be revisited to no longer pass around the System.IO.Stream object
                // from the client side to the server side but instead create a new stream object on the server
                // side and copy the client stream into it. This will require to handle the queuing of multiple
                // successive frames for a given stream.
                Debug.Assert(fin);
                _streamable = frame;

                // If the signal got consumed by another thread in the meantime, make sure to signal the stream to
                // ensure the ReceiveFrameAsync will wake-up and return the frame.
                if (!IsSignaled)
                {
                    SignalCompletion((frame, true));
                }
```

Note that this will need to be revisited to be handle streaming with colocation and not just pass over the IO stream object from client to server side.